retrofit/src/main/java/retrofit/RequestBuilderAction.java

```diff
@@ -199,8 +199,25 @@
     @Override void perform(RequestBuilder builder, Object value) {
       if (value == null) return; // Skip null values.
 
-      //noinspection unchecked
-      builder.addPart(headers, converter.toBody((T) value));
+      if (value instanceof Iterable) {
+        for (Object iterableValue : (Iterable<?>) value) {
+          if (iterableValue != null) { // Skip null values.
+            //noinspection unchecked
+            builder.addPart(headers, converter.toBody((T) iterableValue));
+          }
+        }
+      } else if (value.getClass().isArray()) {
+        for (int x = 0, arrayLength = Array.getLength(value); x < arrayLength; x++) {
+          Object arrayValue = Array.get(value, x);
+          if (arrayValue != null) { // Skip null values.
+            //noinspection unchecked
```

Comment:
Yeah I don't know why these work because I feel like they shouldn't. I'll break apart these behaviors into individual actions so validation happens at configuration time.